src/client/websocket/WebSocketShard.js

```diff
@@ -737,30 +829,104 @@ class WebSocketShard extends EventEmitter {
+     **/
+    if (closeCode === 4009) {
+      this.debug(
+        `[WebSocket] closeCode[4009]: Adding a timeout to check the connection state. | WS State: ${
+          CONNECTION_STATE[this.connection?.readyState ?? WebSocket.CLOSED]
+        }`,
+      );
+      this.setWsCloseTimeout();
+    }
 
-    // Step 4: Cache the old sequence (use to attempt a resume)
-    if (this.sequence !== -1) this.closeSequence = this.sequence;
+    this.debug(
+      `[WS Destroy] Step 2: Resetting the sequence and session id if requested. | WS State: ${
+        CONNECTION_STATE[this.connection?.readyState ?? WebSocket.CLOSED]
+      }`,
+    );
 
-    // Step 5: Reset the sequence and session id if requested
+    // Step 2: Reset the sequence and session id if requested
     if (reset) {
       this.sequence = -1;
       this.sessionId = null;
     }
 
-    // Step 6: reset the rate limit data
+    // Step 3: Watch for zombie connection event which could be emitted when step 1 fails to close the WebSocket
+    this.on(ShardEvents.ZOMBIE_CONNECTION, this.handleZombieConnection);
+
+    this.debug('Step 4: reset the rate limit data');
+    // Step 4: reset the rate limit data
     this.ratelimit.remaining = this.ratelimit.total;
     this.ratelimit.queue.length = 0;
     if (this.ratelimit.timer) {
+      this.debug(`[WS Destroy] Step 4: Clearing ratelimit timer`);
       clearTimeout(this.ratelimit.timer);
       this.ratelimit.timer = null;
     }
   }
 
+  /**
+   * Forcefully closes the WebSocket when the destory() method could not close it normally
+   * using the ws.close(); and ws.terminate();
+   * @private
+   * @returns {Promise<void>}
+   */
+  async handleZombieConnection() {
+    // Continuing from step 2 from the destroy method.
+
+    // Step 3 Check if the connection was closed and readyState is at CLOSED. If not close it.
+    if (this.connection?.readyState === WebSocket.CLOSING || this.connection?.readyState === WebSocket.OPEN) {
+      this.debug(
+        `[WS Destroy] Step 3: WebSocket Still Connected: trying to terminate. | WS State: ${
+          CONNECTION_STATE[this.connection?.readyState ?? WebSocket.CLOSED]
+        }`,
+      );
+
+      // Trying to close WebSocket with terminate..
+      try {
+        this.connection.terminate();
+      } catch {
+        // No op.
+      }
+
+      // Wait for 400ms for ws to close.
+      await sleep(400);
+      this.debug(`[WebSocket] Final Close Check. | WS State: ${CONNECTION_STATE[this.connection?.readyState]}`);
+      if (this.connection?.readyState === WebSocket.CLOSING || this.connection?.readyState === WebSocket.OPEN) {
+        this.debug(
+          // eslint-disable-next-line max-len
+          `[WebSocket] Connection still stuck at Closing: Calling a Manual Destroy of the socket to Close and reconnect. | WS State: ${
+            CONNECTION_STATE[this.connection?.readyState ?? WebSocket.CLOSED]
```

Comment:
Optional chaining is not needed in this scope
```suggestion
      if (this.connection?.readyState === WebSocket.CLOSING || this.connection?.readyState === WebSocket.OPEN) {
        this.debug(
          // eslint-disable-next-line max-len
          `[WebSocket] Connection still stuck at Closing: Calling a Manual Destroy of the socket to Close and reconnect. | WS State: ${
            CONNECTION_STATE[this.connection.readyState ?? WebSocket.CLOSED]
```